src/test/scala/chisel3/tests/AsyncResetRegTest.scala

```diff
@@ -0,0 +1,159 @@
+  var currentValue: BigInt = BigInt(0)
+  var resetValue:   BigInt = BigInt(0)
+  var enable:       Boolean = false
+
+  def getOutput(inputValues: Seq[BigInt], tpe: Type, outputName: String): BigInt = {
+    currentValue
+  }
+
+  override def inputChanged(name: String, value: BigInt): Unit = {
+    name match {
+      case "d"     => nextValue = value
+      case "en"    => enable = value > BigInt(0)
+      case "rst"   =>
+        if(value > BigInt(0)) {
+          nextValue = resetValue
+          currentValue = nextValue
+        }
+      case _ =>
+    }
+    println(s"next $nextValue cur $currentValue, en $enable")
+  }
+
+  override def clockChange(transition: Transition, clockName: String): Unit = {
+    if(transition == PositiveEdge && enable) {
+      currentValue = nextValue
+    }
+  }
+
+  override def outputDependencies(outputName: String): Seq[String] = {
+    Seq("rst", "clk", "d")
+  }
+
+  override def setParams(params: Seq[Param]): Unit = {
+    params.foreach {
+      case firrtl.ir.IntParam("RESET_VALUE", value) =>
+        resetValue = value
+      case _ =>
+      // ignore
+    }
+  }
+}
+
+class AsyncResetBlackBoxFactory extends ScalaBlackBoxFactory {
+  override def createInstance(instanceName: String, blackBoxName: String): Option[ScalaBlackBox] = {
+    blackBoxName match {
+      case "AsyncResetReg" =>
+        Some(add(new AsyncResetRegScalaImpl(instanceName)))
+      case _ =>
+        None
+    }
+  }
+}
+
+class SimpleRegIO(val w: Int) extends Bundle{
+  val d = Input(UInt(w.W))
+  val q = Output(UInt(w.W))
+  val en = Input(Bool())
+}
+
+class AsyncResetRegVec(val w: Int, val init: BigInt) extends Module {
+  val io: SimpleRegIO = IO(new SimpleRegIO(w))
+
+  private val async_regs = List.tabulate(w) { idx =>
+    val on = if (init.testBit(idx)) 1 else 0
+    Module(new AsyncResetReg(on))
+  }
+
+  private val q = for ((reg, idx) <- async_regs.zipWithIndex) yield {
+    reg.io.clk := clock
+    reg.io.rst := reset
+    reg.io.d   := io.d(idx)
+    reg.io.en  := io.en
+    reg.suggestName(s"reg_$idx")
+    reg.io.q
+  }
+
+  //endian
+  io.q := q.reverse.map(_.asUInt()).reduce { (a, b) => Cat(a, b) }
+
+  override def desiredName = s"AsyncResetRegVec_w${w}_i$init"
+
+}
+
+class UsesAsyncResetReg extends Module {
+  val io = IO(new Bundle {
+    val in = Input(UInt(AsyncResetReg.bitWidth.W))
+    val out = Output(UInt(AsyncResetReg.bitWidth.W))
+  })
+
+  val reg = Module(new AsyncResetRegVec(AsyncResetReg.bitWidth, init = 3))
+
+  reg.io.d := io.in
+  reg.clock := clock
+  reg.reset := reset
+  reg.io.en := 1.U
+
+  io.out := reg.io.q
+}
+
+class AsyncResetRegTest extends FreeSpec with ChiselScalatestTester {
+
+  "it should work normally with respect to clock" in {
+
+    val manager = new Testers2OptionsManager {
+      treadleOptions = treadleOptions.copy(
+        blackBoxFactories = Seq(new AsyncResetBlackBoxFactory)
+      )
+    }
+    test(new UsesAsyncResetReg, manager) { c =>
```

Comment:
Overall this seems like a way heavyweight, probably overkill, test. Can you not just instantiate one AsyncResetReg, which is defined as a single-bit register, and directly inline (into `test(...)`) the DUT shim Module? The addition of AsyncResetVec seems unnecessary if the goal is just to determine whether async reset works. We'd hope that whatever underlies the Vec construction would be tested in base Chisel regressions.

I'd also like to get away from the spaghetti-with-meatballs code style that defined the older testers (eg inline Modules and Bundles where there's no re-use - there's no point in making the code reader jump around unnecessarily).

Potentially also should be a TODO somewhere to replace the BlackBox implementation with proper Chisel async reset support when that happens.

Reply:
Broke test into two, one for single bit, one for multi-bit, makes test simpler and clearer.
Multi-bit code uses SingleBit components to keep things DRY